File: android/framework_old/src/main/java/com/dreampany/framework/data/source/pref/BasePref.java
/*
package com.dreampany.frame.data.source.pref;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import com.dreampany.frame.util.AndroidUtil;
import com.github.pwittchen.prefser.library.rx2.Prefser;
import com.github.pwittchen.prefser.library.rx2.TypeToken;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;

*/
/**
 * Created by Hawladar Roman on 6/25/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 *//*

public abstract class BasePref {

    protected Prefser publicPref;
    protected Prefser privatePref;

    protected BasePref(@NonNull Context context) {
        publicPref = new Prefser(context);
        String prefName = getPrivateName(context);
        SharedPreferences pref = context.getSharedPreferences(prefName, Context.MODE_PRIVATE);
        privatePref = new Prefser(pref);
    }

    protected String getPrivateName(Context context) {
        return AndroidUtil.Companion.getPackageName(context);
    }

    public boolean isPubliclyAvailable(String key) {
        return publicPref.contains(key);
    }

    public boolean isPrivateAvailable(String key) {
        return privatePref.contains(key);
    }

    public <T> void setListItem(String key, T item) {
        TypeToken<List<T>> typeToken = new TypeToken<List<T>>() {
        };
        List<T> inputs = new ArrayList<>();
        inputs.add(item);
        List<T> items = privatePref.get(key, typeToken, inputs);
        if (!items.contains(item)) {
            items.add(item);
        }
        privatePref.put(key, items, typeToken);
    }

    public <T> void setListItems(String key, List<T> items) {
        TypeToken<List<T>> typeToken = new TypeToken<List<T>>() {
        };
        privatePref.put(key, items, typeToken);
    }

    public <T> List<T> getListItems(String key) {
        TypeToken<List<T>> typeToken = new TypeToken<List<T>>() {
        };
        List<T> inputs = new ArrayList<>();
        return privatePref.get(key, typeToken, inputs);
    }

    public <T> void setPublicly(String key, T value) {
        publicPref.put(key, value);
    }

    public <T> void setPrivately(String key, T value) {
        privatePref.put(key, value);
    }

    public void setPrivately(String key, String value) {
        privatePref.put(key, value);
    }

    public void removePrivately(String key) {
        privatePref.remove(key);
    }

    public void removePublicly(String key) {
        publicPref.remove(key);
    }

    public void setPublicly(@NonNull String key, String value) {
        publicPref.put(key, value);
    }

    public void setPublicly(@NonNull String key, long value) {
        publicPref.put(key, value);
    }

    public void setPrivately(String key, boolean value) {
        privatePref.put(key, value);
    }

    public void setPrivately(String key, int value) {
        privatePref.put(key, value);
    }

    public void setPrivately(String key, long value) {
        privatePref.put(key, value);
    }

    public <T> T getPublicly(String key, Class<T> classOfT, T defaultValue) {
        return publicPref.get(key, classOfT, defaultValue);
    }

    public <T> T getPrivately(String key, Class<T> classOfT, T defaultValue) {
        return privatePref.get(key, classOfT, defaultValue);
    }

    public String getPublicly(String key, String defaultValue) {
        return publicPref.get(key, String.class, defaultValue);
    }

    public String getPrivately(String key, String defaultValue) {
        return privatePref.get(key, String.class, defaultValue);
    }

    public int getPrivately(String key, int defaultValue) {
        return privatePref.get(key, Integer.class, defaultValue);
    }

    public long getPrivately(String key, long defaultValue) {
        return privatePref.get(key, Long.class, defaultValue);
    }

    public float getPrivately(String key, float defaultValue) {
        return privatePref.get(key, Float.class, defaultValue);
    }

    public boolean getPrivately(String key, boolean defaultValue) {
        return privatePref.get(key, Boolean.class, defaultValue);
    }

    public <T> Flowable<T> observePublic(String key, Class<T> classOfT, T defaultValue) {
        return publicPref.observe(key, classOfT, defaultValue).toFlowable(BackpressureStrategy.LATEST);
    }

    public <T> Flowable<T> observePrivate(String key, Class<T> classOfT, T defaultValue) {
        return privatePref.observe(key, classOfT, defaultValue).toFlowable(BackpressureStrategy.LATEST);
    }

}
*/

File: android/framework_old/src/main/kotlin/com/dreampany/framework/data/source/pref/BasePref.kt
package com.dreampany.framework.data.source.pref

import android.content.Context
import com.dreampany.framework.misc.Constants
import com.github.pwittchen.prefser.library.rx2.Prefser
import com.github.pwittchen.prefser.library.rx2.TypeToken
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable

/**
 * Created by Roman-372 on 7/11/2019
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
abstract class BasePref(val context: Context) {

    private val publicPref: Prefser
    private val privatePref: Prefser

    init {
        publicPref = Prefser(context)
        val prefName = getPrivateName(context)
        val pref = context.getSharedPreferences(prefName, Context.MODE_PRIVATE)
        privatePref = Prefser(pref)
    }

    open fun getPrivateName(context: Context): String {
        return Constants.Pref.DEFAULT
    }

    fun hasPublic(key: String): Boolean {
        return publicPref.contains(key)
    }

    fun hasPrivate(key: String): Boolean {
        return privatePref.contains(key)
    }

    /* setter */
    fun setPublicly(key: String, value: Int) {
        publicPref.put(key, value)
    }

    fun setPublicly(key: String, value: String) {
        publicPref.put(key, value)
    }

    fun <T> setPublicly(key: String, value: T) {
        publicPref.put(key, value)
    }

    fun setPrivately(key: String, value: Int) {
        privatePref.put(key, value)
    }

    fun setPrivately(key: String, value: Long) {
        privatePref.put(key, value)
    }

    fun setPrivately(key: String, value: String) {
        privatePref.put(key, value)
    }

    fun <T> setPrivately(key: String, value: T) {
        privatePref.put(key, value)
    }

    fun getPublicly(key: String, defaultValue: Int): Int {
        return publicPref.preferences.getInt(key, defaultValue)
    }

    fun getPublicly(key: String, defaultValue: Long): Long {
        return publicPref.preferences.getLong(key, defaultValue)
    }

    fun getPublicly(key: String, defaultValue: Boolean): Boolean {
        return publicPref.preferences.getBoolean(key, defaultValue)
    }

    fun getPrivately(key: String, defaultValue: Boolean): Boolean {
        return privatePref.preferences.getBoolean(key, defaultValue)
    }

    fun getPrivately(key: String, defaultValue: Int): Int {
        return privatePref.preferences.getInt(key, defaultValue)
    }


    fun getPrivately(key: String, defaultValue: String): String {
        return privatePref.get(key, String::class.java, defaultValue)
    }


    fun getPrivately(key: String, defaultValue: Long): Long {
        return privatePref.preferences.getLong(key, defaultValue)
    }

    fun getPublicly(key: String, defaultValue: String): String {
        return publicPref.get(key, String::class.java, defaultValue)
    }

    fun <T> getPublicly(key: String, classOfT: Class<T>, defaultValue: T?): T? {
        return publicPref.get(key, classOfT, defaultValue)
    }

    fun <T> getPrivately(key: String, classOfT: Class<T>, defaultValue: T?): T? {
        return privatePref.get(key, classOfT, defaultValue)
    }

    fun <T> getPrivately(key: String, typeTokenOfT: TypeToken<T>, defaultValue: T?): T? {
        return privatePref.get(key, typeTokenOfT, defaultValue)
    }

    fun <T> putPrivately(key: String, value: T) {
        privatePref.put(key, value)
    }

    fun removePrivately(key: String) {
        privatePref.remove(key)
    }

    fun <T> observePublicly(key: String, classOfT: Class<T>, defaultValue: T): Flowable<T> {
        return publicPref.observe(key, classOfT, defaultValue)
            .toFlowable(BackpressureStrategy.LATEST)
    }

    fun <T> observePrivately(key: String, classOfT: Class<T>, defaultValue: T): Flowable<T> {
        return privatePref.observe(key, classOfT, defaultValue)
            .toFlowable(BackpressureStrategy.LATEST)
    }
}